File: src/utils/notificacionesTareas.js
import { Alert, Linking, Platform } from 'react-native';
import * as Notifications from 'expo-notifications';
import storage from '../config/storage';

const CANAL_TAREAS = 'tareas';

Notifications.setNotificationHandler({
  handleNotification: async () => ({
    shouldShowAlert: true,
    shouldShowBanner: true,
    shouldShowList: true,
    shouldPlaySound: true,
    shouldSetBadge: false,
  }),
});

const limpiarTexto = (valor) => {
  if (valor === null || valor === undefined) {
    return '';
  }

  return String(valor).trim();
};

const crearCanalAndroid = async () => {
  if (Platform.OS !== 'android') {
    return;
  }

  await Notifications.setNotificationChannelAsync(CANAL_TAREAS, {
    name: 'Tareas',
    importance: Notifications.AndroidImportance.MAX,
    vibrationPattern: [0, 250, 250, 250],
    lightColor: '#5B4BF2',
    sound: 'default',
    lockscreenVisibility: Notifications.AndroidNotificationVisibility.PUBLIC,
  });
};

export const inicializarNotificaciones = async () => {
  try {
    if (Platform.OS === 'web') {
      console.log('Las notificaciones locales no funcionan en web.');
      return false;
    }

    await crearCanalAndroid();

    const permisosActuales = await Notifications.getPermissionsAsync();

    console.log('Permisos actuales de notificaciones:', permisosActuales);

    let estadoFinal = permisosActuales.status;

    if (estadoFinal !== 'granted') {
      const permisosSolicitados = await Notifications.requestPermissionsAsync();
      console.log('Permisos solicitados:', permisosSolicitados);
      estadoFinal = permisosSolicitados.status;
    }

    if (estadoFinal !== 'granted') {
      Alert.alert(
        'Permiso de notificaciones',
        'No se autorizó el permiso. Actívalo manualmente en la configuración del teléfono.',
        [
          {
            text: 'Abrir configuración',
            onPress: () => Linking.openSettings(),
          },
          {
            text: 'Cancelar',
            style: 'cancel',
          },
        ]
      );

      return false;
    }

    return true;
  } catch (error) {
    console.log('Error inicializando notificaciones:', error);
    return false;
  }
};

const convertirFechaLocal = (valor) => {
  if (!valor) {
    return null;
  }

  if (valor instanceof Date) {
    if (Number.isNaN(valor.getTime())) {
      return null;
    }

    return valor;
  }

  const textoOriginal = String(valor).trim();

  if (!textoOriginal) {
    return null;
  }

  const textoLimpio = textoOriginal
    .replace('T', ' ')
    .replace(/\.\d+/, '')
    .replace(/Z$/i, '')
    .replace(/([+-]\d{2}:?\d{2})$/, '')
    .trim();

  const match = textoLimpio.match(
    /^(\d{4})-(\d{2})-(\d{2})(?:\s+(\d{1,2}):(\d{2})(?::(\d{2}))?)?$/
  );

  if (!match) {
    const fechaFallback = new Date(textoOriginal);

    if (Number.isNaN(fechaFallback.getTime())) {
      return null;
    }

    return fechaFallback;
  }

  const year = Number(match[1]);
  const month = Number(match[2]) - 1;
  const day = Number(match[3]);
  const hours = Number(match[4] || 0);
  const minutes = Number(match[5] || 0);
  const seconds = Number(match[6] || 0);

  const fecha = new Date(year, month, day, hours, minutes, seconds, 0);

  if (Number.isNaN(fecha.getTime())) {
    return null;
  }

  return fecha;
};

const obtenerIdTarea = (tarea) => {
  return limpiarTexto(
    tarea?.id ||
      tarea?.id_tarea ||
      tarea?.idActividad ||
      tarea?.id_actividad ||
      tarea?._id ||
      tarea?.folio ||
      tarea?.nombre ||
      tarea?.titulo ||
      `tarea-${Date.now()}`
  );
};

const obtenerTituloTarea = (tarea) => {
  return limpiarTexto(
    tarea?.titulo ||
      tarea?.nombre ||
      tarea?.actividad ||
      tarea?.descripcion ||
      'Tarea'
  ) || 'Tarea';
};

const obtenerFechaExpiracion = (tarea) => {
  return convertirFechaLocal(
    tarea?.fechaExpiracion ||
      tarea?.fecha_expiracion ||
      tarea?.fechaVencimiento ||
      tarea?.fecha_vencimiento ||
      tarea?.fechaFin ||
      tarea?.fecha_fin ||
      tarea?.vence ||
      tarea?.deadline ||
      tarea?.expira
  );
};

const eliminarStorageItem = async (key) => {
  try {
    if (storage.removeItem) {
      await storage.removeItem(key);
      return;
    }

    if (storage.deleteItem) {
      await storage.deleteItem(key);
      return;
    }

    await storage.setItem(key, '');
  } catch (error) {
    console.log(`Error eliminando ${key}:`, error.message);
  }
};

export const probarNotificacionLocal = async () => {
  try {
    const tienePermiso = await inicializarNotificaciones();

    if (!tienePermiso) {
      console.log('No se pudo probar notificación porque no hay permiso.');
      return null;
    }

    const id = await Notifications.scheduleNotificationAsync({
      content: {
        title: 'Prueba de notificación ✅',
        body: 'Si ves esto, las notificaciones locales ya funcionan.',
        sound: 'default',
        data: {
          tipo: 'prueba_notificacion',
        },
      },
      trigger: {
        type: Notifications.SchedulableTriggerInputTypes.TIME_INTERVAL,
        seconds: 2,
        channelId: CANAL_TAREAS,
      },
    });

    console.log('Notificación de prueba programada:', id);

    return id;
  } catch (error) {
    console.log('Error probando notificación local:', error);
    return null;
  }
};

export const cancelarNotificacionesTarea = async (idTarea) => {
  try {
    const idLimpio = limpiarTexto(idTarea);

    if (!idLimpio) {
      return;
    }

    const key = `notificaciones_tarea_${idLimpio}`;
    const guardadas = await storage.getItem(key);

    if (!guardadas) {
      return;
    }

    const ids = JSON.parse(guardadas);

    if (Array.isArray(ids)) {
      for (const id of ids) {
        if (id) {
          await Notifications.cancelScheduledNotificationAsync(id);
        }
      }
    }

    await eliminarStorageItem(key);
  } catch (error) {
    console.log('Error cancelando notificaciones de tarea:', error);
  }
};

export const programarNotificacionesTarea = async (tarea) => {
  try {
    const tienePermiso = await inicializarNotificaciones();

    if (!tienePermiso) {
      return {
        ok: false,
        ids: [],
        mensaje: 'No hay permiso para notificaciones.',
      };
    }

    const idTarea = obtenerIdTarea(tarea);
    const tituloTarea = obtenerTituloTarea(tarea);
    const fechaExpiracion = obtenerFechaExpiracion(tarea);

    await cancelarNotificacionesTarea(idTarea);

    const idsProgramados = [];
    const ahora = new Date();

    const idHabilitada = await Notifications.scheduleNotificationAsync({
      content: {
        title: 'Tarea habilitada ✅',
        body: `${tituloTarea} ya está activa.`,
        sound: 'default',
        data: {
          tipo: 'tarea_habilitada',
          idTarea,
        },
      },
      trigger: {
        type: Notifications.SchedulableTriggerInputTypes.TIME_INTERVAL,
        seconds: 2,
        channelId: CANAL_TAREAS,
      },
    });

    idsProgramados.push(idHabilitada);

    if (fechaExpiracion) {
      const tiempoExpiracion = fechaExpiracion.getTime();
      const diezMinutosAntes = new Date(tiempoExpiracion - 10 * 60 * 1000);

      if (diezMinutosAntes > ahora) {
        const idDiezMinutos = await Notifications.scheduleNotificationAsync({
          content: {
            title: 'Tu tarea está por expirar ⏳',
            body: `${tituloTarea} vence en 10 minutos.`,
            sound: 'default',
            data: {
              tipo: 'tarea_10_minutos_antes',
              idTarea,
            },
          },
          trigger: {
            type: Notifications.SchedulableTriggerInputTypes.DATE,
            date: diezMinutosAntes,
            channelId: CANAL_TAREAS,
          },
        });

        idsProgramados.push(idDiezMinutos);
      }

      if (fechaExpiracion > ahora) {
        const idExpirada = await Notifications.scheduleNotificationAsync({
          content: {
            title: 'Tarea expirada ⚠️',
            body: `${tituloTarea} ya expiró.`,
            sound: 'default',
            data: {
              tipo: 'tarea_expirada',
              idTarea,
            },
          },
          trigger: {
            type: Notifications.SchedulableTriggerInputTypes.DATE,
            date: fechaExpiracion,
            channelId: CANAL_TAREAS,
          },
        });

        idsProgramados.push(idExpirada);
      }
    }

    await storage.setItem(
      `notificaciones_tarea_${idTarea}`,
      JSON.stringify(idsProgramados)
    );

    console.log('Notificaciones programadas:', idsProgramados);

    return {
      ok: true,
      ids: idsProgramados,
      mensaje: 'Notificaciones programadas correctamente.',
    };
  } catch (error) {
    console.log('Error programando notificaciones de tarea:', error);

    return {
      ok: false,
      ids: [],
      mensaje: error.message,
    };
  }
};

File: lib/utils/notificaciones_tareas.dart
import 'dart:convert';
import 'package:flutter/foundation.dart';
import 'package:flutter/material.dart';
import 'package:flutter_local_notifications/flutter_local_notifications.dart';
import 'package:timezone/data/latest_all.dart' as tz;
import 'package:timezone/timezone.dart' as tz;
import '../config/storage_service.dart';

class NotificacionesTareas {
  static const String canalTareas = 'tareas';

  static final FlutterLocalNotificationsPlugin _plugin =
      FlutterLocalNotificationsPlugin();

  static bool _inicializado = false;

  static Future<bool> inicializarNotificaciones() async {
    try {
      if (kIsWeb) {
        debugPrint('Las notificaciones locales no funcionan en web.');
        return false;
      }

      if (!_inicializado) {
        tz.initializeTimeZones();
        tz.setLocalLocation(tz.getLocation('America/Mexico_City'));

        const androidSettings = AndroidInitializationSettings(
          '@mipmap/ic_launcher',
        );

        const iosSettings = DarwinInitializationSettings(
          requestAlertPermission: true,
          requestBadgePermission: false,
          requestSoundPermission: true,
          defaultPresentAlert: true,
          defaultPresentSound: true,
          defaultPresentBadge: false,
          defaultPresentBanner: true,
          defaultPresentList: true,
        );

        const initSettings = InitializationSettings(
          android: androidSettings,
          iOS: iosSettings,
        );

        await _plugin.initialize(
          settings: initSettings,
          onDidReceiveNotificationResponse: (NotificationResponse response) {
            debugPrint('Notificación presionada: ${response.payload}');
          },
        );

        await _crearCanalAndroid();

        _inicializado = true;
      }

      final permisoAndroid = await _solicitarPermisoAndroid();
      final permisoIos = await _solicitarPermisoIos();

      return permisoAndroid && permisoIos;
    } catch (error) {
      debugPrint('Error inicializando notificaciones: $error');
      return false;
    }
  }

  static Future<void> _crearCanalAndroid() async {
    final androidPlugin =
        _plugin.resolvePlatformSpecificImplementation<
            AndroidFlutterLocalNotificationsPlugin>();

    if (androidPlugin == null) {
      return;
    }

    const channel = AndroidNotificationChannel(
      canalTareas,
      'Tareas',
      description: 'Notificaciones de tareas y misiones',
      importance: Importance.max,
      playSound: true,
      enableVibration: true,
      ledColor: Color(0xFF5B4BF2),
    );

    await androidPlugin.createNotificationChannel(channel);
  }

  static Future<bool> _solicitarPermisoAndroid() async {
    final androidPlugin =
        _plugin.resolvePlatformSpecificImplementation<
            AndroidFlutterLocalNotificationsPlugin>();

    if (androidPlugin == null) {
      return true;
    }

    final permiso = await androidPlugin.requestNotificationsPermission();

    return permiso ?? true;
  }

  static Future<bool> _solicitarPermisoIos() async {
    final iosPlugin =
        _plugin.resolvePlatformSpecificImplementation<
            IOSFlutterLocalNotificationsPlugin>();

    if (iosPlugin == null) {
      return true;
    }

    final permiso = await iosPlugin.requestPermissions(
      alert: true,
      badge: false,
      sound: true,
    );

    return permiso ?? false;
  }

  static NotificationDetails _detallesNotificacion() {
    const androidDetails = AndroidNotificationDetails(
      canalTareas,
      'Tareas',
      channelDescription: 'Notificaciones de tareas y misiones',
      importance: Importance.max,
      priority: Priority.max,
      playSound: true,
      enableVibration: true,
      visibility: NotificationVisibility.public,
    );

    const iosDetails = DarwinNotificationDetails(
      presentAlert: true,
      presentSound: true,
      presentBadge: false,
      presentBanner: true,
      presentList: true,
      interruptionLevel: InterruptionLevel.active,
    );

    return const NotificationDetails(
      android: androidDetails,
      iOS: iosDetails,
    );
  }

  static String _limpiarTexto(dynamic valor) {
    if (valor == null) {
      return '';
    }

    return valor.toString().trim();
  }

  static DateTime? convertirFechaLocal(dynamic valor) {
    if (valor == null) {
      return null;
    }

    if (valor is DateTime) {
      return valor;
    }

    final textoOriginal = valor.toString().trim();

    if (textoOriginal.isEmpty) {
      return null;
    }

    final textoLimpio = textoOriginal
        .replaceAll('T', ' ')
        .replaceAll(RegExp(r'\.\d+'), '')
        .replaceAll(RegExp(r'Z$', caseSensitive: false), '')
        .replaceAll(RegExp(r'([+-]\d{2}:?\d{2})$'), '')
        .trim();

    final regex = RegExp(
      r'^(\d{4})-(\d{2})-(\d{2})(?:\s+(\d{1,2}):(\d{2})(?::(\d{2}))?)?$',
    );

    final match = regex.firstMatch(textoLimpio);

    if (match == null) {
      return DateTime.tryParse(textoOriginal);
    }

    final year = int.parse(match.group(1)!);
    final month = int.parse(match.group(2)!);
    final day = int.parse(match.group(3)!);
    final hours = int.tryParse(match.group(4) ?? '0') ?? 0;
    final minutes = int.tryParse(match.group(5) ?? '0') ?? 0;
    final seconds = int.tryParse(match.group(6) ?? '0') ?? 0;

    return DateTime(
      year,
      month,
      day,
      hours,
      minutes,
      seconds,
    );
  }

  static String obtenerIdTarea(Map<String, dynamic> tarea) {
    return _limpiarTexto(
      tarea['id'] ??
          tarea['id_tarea'] ??
          tarea['idActividad'] ??
          tarea['id_actividad'] ??
          tarea['_id'] ??
          tarea['folio'] ??
          tarea['nombre'] ??
          tarea['titulo'] ??
          'tarea-${DateTime.now().millisecondsSinceEpoch}',
    );
  }

  static String obtenerTituloTarea(Map<String, dynamic> tarea) {
    final titulo = _limpiarTexto(
      tarea['titulo'] ??
          tarea['nombre'] ??
          tarea['actividad'] ??
          tarea['descripcion'] ??
          'Tarea',
    );

    if (titulo.isEmpty) {
      return 'Tarea';
    }

    return titulo;
  }

  static DateTime? obtenerFechaInicio(Map<String, dynamic> tarea) {
    return convertirFechaLocal(
      tarea['fechaInicioNotificacion'] ??
          tarea['fechaInicio'] ??
          tarea['fecha_inicio'] ??
          tarea['fechaHabilitacion'] ??
          tarea['fecha_habilitacion'] ??
          tarea['inicio'] ??
          tarea['empieza'],
    );
  }

  static DateTime? obtenerFechaExpiracion(Map<String, dynamic> tarea) {
    return convertirFechaLocal(
      tarea['fechaExpiracion'] ??
          tarea['fecha_expiracion'] ??
          tarea['fechaLimiteCumplimiento'] ??
          tarea['fecha_limite_cumplimiento'] ??
          tarea['fechaVencimiento'] ??
          tarea['fecha_vencimiento'] ??
          tarea['fechaFin'] ??
          tarea['fecha_fin'] ??
          tarea['vence'] ??
          tarea['deadline'] ??
          tarea['expira'],
    );
  }

  static int _generarIdNotificacion(
    String idTarea,
    String tipo,
  ) {
    final texto = '$idTarea-$tipo';
    int hash = 0;

    for (int i = 0; i < texto.length; i++) {
      hash = 0x1fffffff & (hash + texto.codeUnitAt(i));
      hash = 0x1fffffff & (hash + ((0x0007ffff & hash) << 10));
      hash = hash ^ (hash >> 6);
    }

    hash = 0x1fffffff & (hash + ((0x03ffffff & hash) << 3));
    hash = hash ^ (hash >> 11);
    hash = 0x1fffffff & (hash + ((0x00003fff & hash) << 15));

    if (hash == 0) {
      return DateTime.now().millisecondsSinceEpoch.remainder(2147483647);
    }

    return hash;
  }

  static tz.TZDateTime _toTzDateTime(DateTime fecha) {
    return tz.TZDateTime.from(fecha, tz.local);
  }

  static DateTime? _fechaFuturaSegura(
    DateTime? fecha, {
    int segundosMinimos = 5,
  }) {
    if (fecha == null) {
      return null;
    }

    final ahora = DateTime.now();
    final minimo = ahora.add(Duration(seconds: segundosMinimos));

    if (fecha.isBefore(minimo) || fecha.isAtSameMomentAs(minimo)) {
      return null;
    }

    return fecha;
  }

  static Future<void> _programarNotificacion({
    required int id,
    required String title,
    required String body,
    required DateTime fecha,
    required Map<String, dynamic> payload,
  }) async {
    await _plugin.zonedSchedule(
      id: id,
      title: title,
      body: body,
      scheduledDate: _toTzDateTime(fecha),
      notificationDetails: _detallesNotificacion(),
      androidScheduleMode: AndroidScheduleMode.inexactAllowWhileIdle,
      payload: jsonEncode(payload),
    );
  }

  static Future<int?> probarNotificacionInmediataIos() async {
    try {
      final tienePermiso = await inicializarNotificaciones();

      debugPrint('Permiso notificaciones iOS inmediata: $tienePermiso');

      if (!tienePermiso) {
        debugPrint('No hay permiso para mostrar notificaciones.');
        return null;
      }

      final id = _generarIdNotificacion(
        'prueba-inmediata-ios',
        DateTime.now().millisecondsSinceEpoch.toString(),
      );

      await _plugin.show(
        id: id,
        title: 'Prueba inmediata ✅',
        body: 'Si ves esto, iOS ya permite notificaciones locales.',
        notificationDetails: _detallesNotificacion(),
        payload: jsonEncode({
          'tipo': 'prueba_inmediata_ios',
        }),
      );

      debugPrint('Notificación inmediata enviada: $id');

      return id;
    } catch (error) {
      debugPrint('Error enviando notificación inmediata iOS: $error');
      return null;
    }
  }

  static Future<int?> probarNotificacionProgramadaIos() async {
    try {
      final tienePermiso = await inicializarNotificaciones();

      debugPrint('Permiso notificaciones iOS programada: $tienePermiso');

      if (!tienePermiso) {
        debugPrint('No hay permiso para programar notificaciones.');
        return null;
      }

      final id = _generarIdNotificacion(
        'prueba-programada-ios',
        DateTime.now().millisecondsSinceEpoch.toString(),
      );

      final fechaProgramada = DateTime.now().add(
        const Duration(seconds: 20),
      );

      await _programarNotificacion(
        id: id,
        title: 'Prueba programada ⏰',
        body: 'Esta notificación estaba programada, no salió por el botón.',
        fecha: fechaProgramada,
        payload: {
          'tipo': 'prueba_programada_ios',
        },
      );

      debugPrint('Notificación programada iOS: $id');
      debugPrint('Fecha programada: $fechaProgramada');

      return id;
    } catch (error) {
      debugPrint('Error programando notificación iOS: $error');
      return null;
    }
  }

  static Future<int?> probarNotificacionLocal() async {
    return probarNotificacionProgramadaIos();
  }

  static Future<void> cancelarNotificacionesTarea(dynamic idTarea) async {
    try {
      final idLimpio = _limpiarTexto(idTarea);

      if (idLimpio.isEmpty) {
        return;
      }

      final key = 'notificaciones_tarea_$idLimpio';
      final guardadas = await StorageService.getItem(key);

      if (guardadas == null || guardadas.isEmpty) {
        return;
      }

      final ids = jsonDecode(guardadas);

      if (ids is List) {
        for (final id in ids) {
          final idInt = int.tryParse(id.toString());

          if (idInt != null) {
            await _plugin.cancel(id: idInt);
          }
        }
      }

      await StorageService.removeItem(key);
    } catch (error) {
      debugPrint('Error cancelando notificaciones de tarea: $error');
    }
  }

  static Future<Map<String, dynamic>> programarNotificacionesTarea(
    Map<String, dynamic> tarea,
  ) async {
    try {
      final tienePermiso = await inicializarNotificaciones();

      if (!tienePermiso) {
        return {
          'ok': false,
          'ids': <int>[],
          'mensaje': 'No hay permiso para notificaciones.',
        };
      }

      final idTarea = obtenerIdTarea(tarea);
      final tituloTarea = obtenerTituloTarea(tarea);
      final fechaInicio = obtenerFechaInicio(tarea);
      final fechaExpiracion = obtenerFechaExpiracion(tarea);

      await cancelarNotificacionesTarea(idTarea);

      final idsProgramados = <int>[];

      final fechaInicioSegura = _fechaFuturaSegura(
        fechaInicio,
        segundosMinimos: 5,
      );

      if (fechaInicioSegura != null) {
        final idHabilitada = _generarIdNotificacion(
          idTarea,
          'habilitada',
        );

        await _programarNotificacion(
          id: idHabilitada,
          title: 'Tarea habilitada ✅',
          body: '$tituloTarea ya está activa.',
          fecha: fechaInicioSegura,
          payload: {
            'tipo': 'tarea_habilitada',
            'idTarea': idTarea,
          },
        );

        idsProgramados.add(idHabilitada);
      }

      if (fechaExpiracion != null) {
        final diezMinutosAntes = fechaExpiracion.subtract(
          const Duration(minutes: 10),
        );

        final fechaDiezMinutosSegura = _fechaFuturaSegura(
          diezMinutosAntes,
          segundosMinimos: 5,
        );

        if (fechaDiezMinutosSegura != null) {
          final idDiezMinutos = _generarIdNotificacion(
            idTarea,
            '10_minutos_antes',
          );

          await _programarNotificacion(
            id: idDiezMinutos,
            title: 'Tu tarea está por expirar ⏳',
            body: '$tituloTarea vence en 10 minutos.',
            fecha: fechaDiezMinutosSegura,
            payload: {
              'tipo': 'tarea_10_minutos_antes',
              'idTarea': idTarea,
            },
          );

          idsProgramados.add(idDiezMinutos);
        }

        final fechaExpiracionSegura = _fechaFuturaSegura(
          fechaExpiracion,
          segundosMinimos: 5,
        );

        if (fechaExpiracionSegura != null) {
          final idExpirada = _generarIdNotificacion(
            idTarea,
            'expirada',
          );

          await _programarNotificacion(
            id: idExpirada,
            title: 'Tarea expirada ⚠️',
            body: '$tituloTarea ya expiró.',
            fecha: fechaExpiracionSegura,
            payload: {
              'tipo': 'tarea_expirada',
              'idTarea': idTarea,
            },
          );

          idsProgramados.add(idExpirada);
        }
      }

      await StorageService.setItem(
        'notificaciones_tarea_$idTarea',
        jsonEncode(idsProgramados),
      );

      debugPrint('Notificaciones programadas para tarea $idTarea: $idsProgramados');
      debugPrint('Fecha inicio detectada: $fechaInicio');
      debugPrint('Fecha expiración detectada: $fechaExpiracion');

      return {
        'ok': true,
        'ids': idsProgramados,
        'mensaje': idsProgramados.isEmpty
            ? 'No había fechas futuras para programar.'
            : 'Notificaciones programadas correctamente.',
      };
    } catch (error) {
      debugPrint('Error programando notificaciones de tarea: $error');

      return {
        'ok': false,
        'ids': <int>[],
        'mensaje': error.toString(),
      };
    }
  }
}